Split Parser into command-specific classes #137

MightyCupcakes opened the pull request:
Closes #137 

This PR continues from #247 

This PR will adopt the simpliest method of implementing #137 which is to split all parsing code for different commands into their own classes. Commands like List, Exit and Clear does not perform any parsing and thus are not affected.

CanIHasReview commented:
# v1

@MightyCupcakes submitted v1 for review.

* [v1 1/6] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/cb0848bd86a3dd32fa5793db83aeef798349a4ab)
* [v1 2/6] [Extract AddCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/ea29fda6c1448e07c7d4c392a2e9a4fb0288db89)
* [v1 3/6] [Extract EditCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/b64144f3e7b21e17635691e01bdd1996db5fea7a)
* [v1 4/6] [Extract DeleteCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/e9a8af01ead37cdb77ce3471c47c938ad08af3be)
* [v1 5/6] [Extract SelectCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/8fdb97e0fc8632fc66bae8146417a524a4fad382)
* [v1 6/6] [Extract FindCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/8da3a69831c9b74fa50e48cf5ffc6af0d0d4e153)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/da4dc8e78218b35ccf991108acf1c56ca8492f57...8da3a69831c9b74fa50e48cf5ffc6af0d0d4e153))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

yamgent commented on src/main/java/seedu/address/logic/parser/ParserUtil.java:
```diff
@@ -0,0 +1,47 @@
+import seedu.address.commons.util.StringUtil;
+
+/**
+ * A container for utility functions used in parsing
```
I don't think you need to state `ParserUtil` as "a container". It sounds weird, especially when other utility classes in addressbook do not refer themselves as containers, even though they also provide utility methods like `ParserUtil`.

You can just say `Utility methods used for parsing strings in the various *Parser classes`.

damithc commented on src/main/java/seedu/address/logic/parser/ParserUtil.java:
```diff
@@ -0,0 +1,47 @@
+    public static Set<String> toTagSet(Optional<List<String>> tagsOptional) {
+        List<String> tags = tagsOptional.orElse(Collections.emptyList());
+        return new HashSet<>(tags);
+    }
```
This method doesn't seem to be specific to tags although the code uses `tags` in method/parameter names.

chao1995 commented on src/main/java/seedu/address/logic/parser/EditCommandParser.java:
```diff
@@ -0,0 +1,139 @@
+    * @return A list of size {@code numFields} where the ith element is the ith field value if specified in
+    *         the input, {@code Optional.empty()} otherwise.
+    */
+    private List<Optional<String>> splitPreamble(String preamble, int numFields) {
```
I think this method and other parseSomething functions can be moved to ParserUtil.

MightyCupcakes commented:
@chao1995 

Regarding the commit message - yes newcomers wont know about it but I think it is important because this point was raised up during a discussion and reviewing process and would be lost if it is in included in the commit.

I will add in your point about SRP in the commit message though as it will provide more context for new developers

CanIHasReview commented:
# v2

@MightyCupcakes submitted v2 for review.

* [v2 1/6] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/1f618f629f577630cb9cc53a942312b646e7976f)
* [v2 2/6] [Extract AddCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/532750fdc2903fbedb8d54b5d360de22a8fa3a01)
* [v2 3/6] [Extract EditCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/2ee7b2d87cdbf2fff4ec7aca405c5eabb6d07047)
* [v2 4/6] [Extract DeleteCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/a0ff0db18b51cb69601aed2df5ec8765290e099f)
* [v2 5/6] [Extract SelectCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/728d89da20cdb1b386b01cc3edab91545dc97420)
* [v2 6/6] [Extract FindCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/4654114ee86c6f62e1b183d15ca6c1bd1a665beb)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/da4dc8e78218b35ccf991108acf1c56ca8492f57...4654114ee86c6f62e1b183d15ca6c1bd1a665beb)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/commit/1e63ef75c150c09fa121f5a687d5be5b3cd9cfa0))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

yamgent commented on src/main/java/seedu/address/logic/parser/ParserUtil.java:
```diff
@@ -0,0 +1,111 @@
+     * or if the list contained in the {@code Optional} is empty
+     */
+    public static Set<String> listOptionalToSet(Optional<List<String>> tagsOptional) {
+        List<String> tags = tagsOptional.orElse(Collections.emptyList());
```
The method has been renamed, but the parameters and variables are still referring to tags.

Also, should this be in `StringUtil` instead?

MightyCupcakes replied:
Changes in #261 will render this function useless and removed by the same PR. Therefore I am really at loss what should I do with this function. It is really meant to be a private utility function abstracting out the conversion of an Optional value from `ArgumentsTokenizer` into a Set of tags so its place in `StringUtil` is definitely not justified.

damithc replied:
It does seem to have a very specific narrow scope job. Can keep it here. However, get rid of the `tag` references in the code for the time being? Also, `listOptional` is implied by the parameter; name it just `toSet`

```java
public static Set<String> toSet(Optional<List<String>> list) {
    List<String> elements = list.orElse(Collections.emptyList());
    return new HashSet<>(elements);
}
```

damithc commented on src/main/java/seedu/address/logic/parser/EditCommandParser.java:
```diff
@@ -0,0 +1,75 @@
+            editPersonDescriptor.setEmail(ParserUtil.parseEmail(argsTokenizer.getValue(PREFIX_EMAIL)));
+            editPersonDescriptor.setAddress(ParserUtil.parseAddress(argsTokenizer.getValue(PREFIX_ADDRESS)));
+            editPersonDescriptor.setTags(parseTagsForEdit(
+                    ParserUtil.listOptionalToSet(argsTokenizer.getAllValues(PREFIX_TAG))));
```
Prefer higher level line breaks to lower level ones. e.g. 
```java
editPersonDescriptor.setTags(
        parseTagsForEdit(ParserUtil.listOptionalToSet(argsTokenizer.getAllValues(PREFIX_TAG))));
```
or 
```java
editPersonDescriptor
        .setTags(parseTagsForEdit(ParserUtil.listOptionalToSet(argsTokenizer.getAllValues(PREFIX_TAG))));
```

damithc commented on src/main/java/seedu/address/logic/parser/FindCommandParser.java:
```diff
@@ -0,0 +1,36 @@
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
+        }
+
+        // keywords delimited by whitespace
```
If this is only applicable to line 32, should be given as a trailing comment in the same line or there should be a blank line after line 32

damithc commented:
@chao1995 any input for this iteration?

CanIHasReview commented:
# v3

@MightyCupcakes submitted v3 for review.

* [v3 1/6] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/35020ccffd430fee55f6696f9883e90244458482)
* [v3 2/6] [Extract AddCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/ce9f24b71925ea460d9beb617dee0e6f390f3824)
* [v3 3/6] [Extract EditCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/d41a09f98cef7ce4690ce12ca033e9080dc952bd)
* [v3 4/6] [Extract DeleteCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/e9fc5fea1adb075d6cee23feae4a3c31f5986c2a)
* [v3 5/6] [Extract SelectCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/223a236497465de29ae69ef9985787e77534e7f8)
* [v3 6/6] [Extract FindCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/1a8e6e3c6264a2a93e56507f90f51a69b9d85a79)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/a6c0c981f15005501734ffde9c310dd0d3be2699...1a8e6e3c6264a2a93e56507f90f51a69b9d85a79)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/commit/766887d04817cb0f4d8542817ebf40105065f256))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

damithc commented:
@yamgent @chao1995 let's review fast. This is a high priority PR.

damithc commented:
@MightyCupcakes you need to update docs in the next iteration. Otherwise docs will not match the code when this PR is merged.

chao1995 commented on src/main/java/seedu/address/logic/parser/CliSyntax.java:
```diff
@@ -0,0 +1,29 @@
+
+/**
+ * A container class to contain all general Command Line Interface (CLI)
+ * parsing related syntax definitions
```
> I wonder if we can make this an interface called CliParser and let all CommandParser classes implement it

Using interface to hold constants is generally considered a poor use of interface as described in [this answer](http://stackoverflow.com/a/320642/3522482).

To summarise, the reason is mainly that a class using some constants internally is an implementation detail, however, implementing a constant interface causes this implementation detail to leak into the class's exported API.

Using an Enum may be a better approach.

damithc replied:
> Using interface to hold constants is generally considered a poor use of interface

That's a valid point.

damithc commented on src/main/java/seedu/address/logic/parser/CliSyntax.java:
```diff
@@ -0,0 +1,28 @@
+    public static final Prefix PREFIX_TAG = new Prefix("t/");
+
+    /* Patterns definitions */
+
```
Remove this blank line?

CanIHasReview commented:
# v4

@MightyCupcakes submitted v4 for review.

* [v4 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/82865dcb008b790223f993422a30480571a85607)
* [v4 2/7] [Extract AddCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/9065d3b5b595b8672bfa31adeca1d09a41b54d17)
* [v4 3/7] [Extract EditCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/1d31fe740141b2bfb6725d7388d3d656c99d2211)
* [v4 4/7] [Extract DeleteCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/1d6bd8c31d2b506d2c3b6e8d567bbe3d1bfafb5d)
* [v4 5/7] [Extract SelectCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/e8d3878a9c2909bef4f08c75c4656a2d58bc243b)
* [v4 6/7] [Extract FindCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/a1c512a1629a3ef3111b8f7c4872d8c81c012f8e)
* [v4 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/522250bd2d2a10b4106ede08f298c09685663445)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/941d0afebe52dc6bd86b60bd4a11861f48273900...522250bd2d2a10b4106ede08f298c09685663445)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/commit/6a53f9daa1fa36b0bf6b124b985028cf3a8e7e55))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

chao1995 commented:
Some feedbacks on the images.

DeletePersonSdForLogic.png

![deletepersonsdforlogic](https://cloud.githubusercontent.com/assets/7572632/22515045/30418df8-e8dc-11e6-8d9f-387f116c718d.png)

LogicClassDiagram.png

![logicclassdiagram](https://cloud.githubusercontent.com/assets/7572632/22515047/31b14980-e8dc-11e6-93d5-6842c93b4969.png)

yamgent commented:
Agreed with changes suggested by @chao1995.

CanIHasReview commented:
# v5

@MightyCupcakes submitted v5 for review.

* [v5 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/82865dcb008b790223f993422a30480571a85607)
* [v5 2/7] [Extract AddCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/9065d3b5b595b8672bfa31adeca1d09a41b54d17)
* [v5 3/7] [Extract EditCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/1d31fe740141b2bfb6725d7388d3d656c99d2211)
* [v5 4/7] [Extract DeleteCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/1d6bd8c31d2b506d2c3b6e8d567bbe3d1bfafb5d)
* [v5 5/7] [Extract SelectCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/e8d3878a9c2909bef4f08c75c4656a2d58bc243b)
* [v5 6/7] [Extract FindCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/a1c512a1629a3ef3111b8f7c4872d8c81c012f8e)
* [v5 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/5d82b1a83b39ec22f9c055e9791e9356391e325d)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/941d0afebe52dc6bd86b60bd4a11861f48273900...5d82b1a83b39ec22f9c055e9791e9356391e325d)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/commit/5629b4c0c1db7d55c5b404f1142e57f2b17e8181))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

chao1995 commented:
Minor change for LogicClassDiagram.png.

![logicclassdiagram](https://cloud.githubusercontent.com/assets/7572632/22538691/46bb8368-e94f-11e6-8b44-f66c7cb49a28.png)

yamgent commented:
@MightyCupcakes I updated the review, as I realized that the code is actually creating the Command in the `parse()` method, not in the constructor method. Your sequence diagram need to show that more clearly.

CanIHasReview commented:
# v6

@MightyCupcakes submitted v6 for review.

* [v6 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/82865dcb008b790223f993422a30480571a85607)
* [v6 2/7] [Extract AddCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/9065d3b5b595b8672bfa31adeca1d09a41b54d17)
* [v6 3/7] [Extract EditCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/1d31fe740141b2bfb6725d7388d3d656c99d2211)
* [v6 4/7] [Extract DeleteCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/1d6bd8c31d2b506d2c3b6e8d567bbe3d1bfafb5d)
* [v6 5/7] [Extract SelectCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/e8d3878a9c2909bef4f08c75c4656a2d58bc243b)
* [v6 6/7] [Extract FindCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/a1c512a1629a3ef3111b8f7c4872d8c81c012f8e)
* [v6 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/2fc3617aff68d4a49b4384766f28bde45fafdc03)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/941d0afebe52dc6bd86b60bd4a11861f48273900...2fc3617aff68d4a49b4384766f28bde45fafdc03)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/commit/18bbd7a52734e306e8b2242a5f6b12c918fba078))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

damithc commented:
I tweaked the diagram a bit. File attached.
![image](https://cloud.githubusercontent.com/assets/1673303/22548137/07e90b7a-e980-11e6-94c3-a3fe0ee68a6f.png)

[LogicComponentClassDiagram.pptx](https://github.com/se-edu/addressbook-level4/files/747586/LogicComponentClassDiagram.pptx)

CanIHasReview commented:
# v7

@MightyCupcakes submitted v7 for review.

* [v7 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/82865dcb008b790223f993422a30480571a85607)
* [v7 2/7] [Extract AddCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/9065d3b5b595b8672bfa31adeca1d09a41b54d17)
* [v7 3/7] [Extract EditCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/1d31fe740141b2bfb6725d7388d3d656c99d2211)
* [v7 4/7] [Extract DeleteCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/1d6bd8c31d2b506d2c3b6e8d567bbe3d1bfafb5d)
* [v7 5/7] [Extract SelectCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/e8d3878a9c2909bef4f08c75c4656a2d58bc243b)
* [v7 6/7] [Extract FindCommandParser from Parser class ](https://github.com/se-edu/addressbook-level4/pull/272/commits/a1c512a1629a3ef3111b8f7c4872d8c81c012f8e)
* [v7 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/73359c80aca3cc304a68208a2880303503a91c3e)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/941d0afebe52dc6bd86b60bd4a11861f48273900...73359c80aca3cc304a68208a2880303503a91c3e)) ([:chart_with_upwards_trend: Interdiff between v6 and v7](https://github.com/se-edu/addressbook-level4/commit/8c4d9479335f31bf882cca8e255324266d1239dd))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/7/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

pyokagan commented on src/main/java/seedu/address/logic/parser/CliSyntax.java:
```diff
@@ -0,0 +1,27 @@
+    public static final Pattern INDEX_ARGS_FORMAT = Pattern.compile("(?<targetIndex>.+)");
+
+    /** Used for initial separation of command word and args. */
+    public static final Pattern BASIC_COMMAND_FORMAT = Pattern.compile("(?<commandWord>\\S+)(?<arguments>.*)");
```
Following the current design, this would only be used in `Parser`, right? Is there a use case where another parser class would need to use this as well?

In other words, should this remain private to `Parser`?

pyokagan commented on src/main/java/seedu/address/logic/parser/CliSyntax.java:
```diff
@@ -0,0 +1,27 @@
+    public static final Prefix PREFIX_TAG = new Prefix("t/");
+
+    /* Patterns definitions */
+    public static final Pattern INDEX_ARGS_FORMAT = Pattern.compile("(?<targetIndex>.+)");
```
Shouldn't parsers be using `parseIndex()` rather than accessing this regex directly?

In other words, should this be private to `ParserUtil`?

pyokagan commented on src/main/java/seedu/address/logic/parser/Parser.java:
```diff
@@ -304,10 +209,8 @@ private Command prepareFind(String args) {
 
-        // keywords delimited by whitespace
-        final String[] keywords = matcher.group("keywords").split("\\s+");
+        final String[] keywords = matcher.group("keywords").split("\\s+"); // keywords delimited by whitespace
```
Unnecessary change?

pyokagan commented on src/main/java/seedu/address/logic/parser/Parser.java:
```diff
@@ -304,10 +209,8 @@ private Command prepareFind(String args) {
         final Set<String> keywordSet = new HashSet<>(Arrays.asList(keywords));
         return new FindCommand(keywordSet);
     }
-
```
Unnecessary change?

pyokagan commented on src/main/java/seedu/address/logic/parser/AddCommandParser.java:
```diff
@@ -0,0 +1,43 @@
+            return new IncorrectCommand(ive.getMessage());
+        }
+    }
+}
```
If we have a blank line after the `public class AddCommandParser {` above, then I guess we should also have a blank line here to make it symmetrical.

CanIHasReview commented:
# v8

@MightyCupcakes submitted v8 for review.

* [v8 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/068046a7876da9969d7bf53635ce8d7e8da3f23b)
* [v8 2/7] [Extract AddCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/4f4715c83de5a534e5323b35d91681defb6db619)
* [v8 3/7] [Extract EditCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/ba28f46f5e9fe5f7cd67fceb4da400e4fedda056)
* [v8 4/7] [Extract DeleteCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/28dc6c768870d38b7fabdb0a24c562c11b3a5da3)
* [v8 5/7] [Extract SelectCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/d9c27d8f0eceda4f90b9f8badffaf9e8e5639b13)
* [v8 6/7] [Extract FindCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/5484a92bec689c04b519da9196a6d3ae3240bb69)
* [v8 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/9fd0e7cbf9ac73b685e27669cb89a4f1c9515442)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/2221d24f2e49a9ec8a9c0d45b9dcb03694d25b0f...9fd0e7cbf9ac73b685e27669cb89a4f1c9515442)) ([:chart_with_upwards_trend: Interdiff between v7 and v8](https://github.com/se-edu/addressbook-level4/compare/8598cf16b2315be4279562a12be901bac61e9db4~1...8598cf16b2315be4279562a12be901bac61e9db4))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/8/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

pyokagan commented:
@chao1995 @yamgent @damithc I guess it's annoying if @CanIHasReview keeps re-requesting reviews every iteration :-P

damithc commented:
@pyokagan 
> I guess it's annoying if @CanIHasReview keeps re-requesting reviews from reviewers who already gave approvals every iteration :-P

Yes, that is a downside. As you suggested on slack, a more reasonable middle ground is to re-request reviews only from those who have not given approval yet and get everyone's approval one more time just before merging.

chao1995 commented:
@pyokagan 

> I guess it's annoying if @CanIHasReview keeps re-requesting reviews from reviewers who already gave approvals every iteration :-P

I actually prefer it requests review from me again (because that means I miss something in my review). So it's not annoying to me.

But perhaps the downside is that the more reviewers you request, the slower the iteration might be.

damithc commented:
@MightyCupcakes I think you can go ahead with the next iteration as rest of the reviewers have approved a previous version of this PR. Especially because we need to merge this urgently.

CanIHasReview commented:
# v9

@MightyCupcakes submitted v9 for review.

* [v9 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/59e3208bad7a6b33752eede19b9416226333958b)
* [v9 2/7] [Extract AddCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/952f1c1fe5abd7265bfb08e06004de89cb91c280)
* [v9 3/7] [Extract EditCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/4e2d91979edd6be4466d5d9e4b5e35a4c7d286ff)
* [v9 4/7] [Extract DeleteCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/a4875835d2aa3d843cbf21b699494fe30d3ef6f4)
* [v9 5/7] [Extract SelectCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/e064cfd3e22f4e082710b70055fabd728ab89bc4)
* [v9 6/7] [Extract FindCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/4d362a85fe1b72c660093b141b2fd84d582eb0de)
* [v9 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/c0f138b97a350011fbd58deca5eceb46745b3620)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/2221d24f2e49a9ec8a9c0d45b9dcb03694d25b0f...c0f138b97a350011fbd58deca5eceb46745b3620)) ([:chart_with_upwards_trend: Interdiff between v8 and v9](https://github.com/se-edu/addressbook-level4/compare/7cd0abc3aedd8f3b5f763c8783b691a77b074cfe~1...7cd0abc3aedd8f3b5f763c8783b691a77b074cfe))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/9/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

damithc commented on src/main/java/seedu/address/logic/parser/AddCommandParser.java:
```diff
@@ -0,0 +1,44 @@
+                    argsTokenizer.getValue(PREFIX_EMAIL).get(),
+                    argsTokenizer.getValue(PREFIX_ADDRESS).get(),
+                    ParserUtil.toSet(argsTokenizer.getAllValues(PREFIX_TAG))
+                    );
```
Um, closing bracket should be in the previous line? This is a long line wrapped into multiple lines (as opposed to a code block).

damithc commented:
@chao1995 any comments?

chao1995 commented:
@damithc No comments from me.

damithc commented:
@MightyCupcakes can proceed to the next iteration

CanIHasReview commented:
# v10

@MightyCupcakes submitted v10 for review.

* [v10 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/bb8b6ddbca0502b4f9b4179f3c8c8a8efef14ec0)
* [v10 2/7] [Extract AddCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/1c9ce3b12a6dbc470ef3137812754cf3529e78c5)
* [v10 3/7] [Extract EditCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/94613b9233b407d4d8acfb3ecaed03173d79c5d5)
* [v10 4/7] [Extract DeleteCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/dc7b8de03a23d5636d19ab03989f752cfa049029)
* [v10 5/7] [Extract SelectCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/a7b88f45e89798ce354e1ededa525440b6472e94)
* [v10 6/7] [Extract FindCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/1da36d50bcad403fbab3ca5897701cedef2c244c)
* [v10 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/2d2d506ca94743b89ba998ec164a46e4b4cfc9fa)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...2d2d506ca94743b89ba998ec164a46e4b4cfc9fa)) ([:chart_with_upwards_trend: Interdiff between v9 and v10](https://github.com/se-edu/addressbook-level4/compare/5e46fb12692677e434272c9b143b13ab682c20bd~1...5e46fb12692677e434272c9b143b13ab682c20bd))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/10/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

MightyCupcakes commented:
# v10

- Updated commit messages of [v10 2/7] to [v10 6/7]. Removed instances of `current` and other unnecessary words.
- Moved closing bracket in `AddCommandParser` to the same line as the last parameter.

pyokagan commented on src/main/java/seedu/address/logic/parser/AddCommandParser.java:
```diff
@@ -0,0 +1,43 @@
+                    argsTokenizer.getValue(PREFIX_PHONE).get(),
+                    argsTokenizer.getValue(PREFIX_EMAIL).get(),
+                    argsTokenizer.getValue(PREFIX_ADDRESS).get(),
+                    ParserUtil.toSet(argsTokenizer.getAllValues(PREFIX_TAG)));
```
Not a direct cut-and-paste from `AddCommandParser.java`, but oh well.

pyokagan replied:
Try not to do these tiny style changes when doing code movement.

pyokagan replied:
> but oh well.

To clarify: my "Oh well" means "This PR is nearly perfect, it is just unfortunate it has all these small micronits". Usually, these micronits would not affect the mergeability of a PR (we need to balance it with other factors such as time). These micronits could even be fixed up on the maintainer's side as well to save a review roundtrip. However, given that CS3281 students would likely be the future stewards of this project, we will need to enforce and aim for a higher level of quality and rigor :-)

pyokagan commented on src/main/java/seedu/address/logic/parser/FindCommandParser.java:
```diff
@@ -0,0 +1,36 @@
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
+        }
+
+        final String[] keywords = matcher.group("keywords").split("\\s+"); // keywords delimited by whitespace
```
Also not a direct cut-and-paste from `Parser.java`, but oh well.

damithc commented:
@MightyCupcakes 
I'm not entirely satisfied with the commit message format. In future, let's try to stick to this flow:

```
{current situation}
{why it is bad}
{what is being done about it}
```

e.g.
```
Parser class contains code for parsing all command types.
This makes the Parser class bloated and can be considered
a violation of the Single Responsibility Principle. The solution
is to extract code for parsing each command into a separate
*CommandParser class. However, that can result in bidirectional 
links between those classes and the Parser class because some
of the constants and utility methods in Parser class will be used
by those classes as well.

Let's extract reusable constants into a new CliSyntax class and 
utility methods into a new ParserUtil class so that such
bidirectional links can be avoided.
```

CanIHasReview commented:
# v11

@MightyCupcakes submitted v11 for review.

* [v11 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/b7fb6b0212d7076f6207688ada49e9cf44e1768f)
* [v11 2/7] [Extract AddCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/f656fee2132510c367bbf5ec282e8eec4e4eebc7)
* [v11 3/7] [Extract EditCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/947c0f582216e29d8f929cb94b048eb754846673)
* [v11 4/7] [Extract DeleteCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/7797d01fcc80bd2c1b1d3b066414c371b8f7d55b)
* [v11 5/7] [Extract SelectCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/951521629e8369749a09c05f8250c2072db7ac93)
* [v11 6/7] [Extract FindCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/444a961ce53022486218f93ca6f7e461e9e4e22d)
* [v11 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/22ff65fc8b0146809b705d46d58ce20a637f94d4)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...22ff65fc8b0146809b705d46d58ce20a637f94d4)) ([:chart_with_upwards_trend: Interdiff between v10 and v11](https://github.com/se-edu/addressbook-level4/compare/4cc7fba8234e16df402a89b9d5956121252015f6~1...4cc7fba8234e16df402a89b9d5956121252015f6))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/11/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

CanIHasReview commented:
# v12

@MightyCupcakes submitted v12 for review.

* [v12 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/b7fb6b0212d7076f6207688ada49e9cf44e1768f)
* [v12 2/7] [Extract AddCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/24eb0480059217a66ac6804008c7fe3d270da483)
* [v12 3/7] [Extract EditCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/e8268d31ddec7771a00268dc831be7d5428fd473)
* [v12 4/7] [Extract DeleteCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/f72f46f82ea8557a46c47a4722c2d5940dbb8a10)
* [v12 5/7] [Extract SelectCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/47cadb701101863f09080337f795b7ba8a76e16d)
* [v12 6/7] [Extract FindCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/1b03213886a0edf9a3915218b6d6d7b8dd426a49)
* [v12 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/998a444cf127da61cf02352350101a048706f446)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...998a444cf127da61cf02352350101a048706f446)) ([:chart_with_upwards_trend: Interdiff between v11 and v12](https://github.com/se-edu/addressbook-level4/compare/dc19e9a8dc025cf8f2beb241284f5db0571a806f~1...dc19e9a8dc025cf8f2beb241284f5db0571a806f))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/12/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

pyokagan commented on src/main/java/seedu/address/logic/parser/Parser.java:
```diff
@@ -29,33 +32,16 @@
-    /**
-     * Used for initial separation of command word and args.
-     */
+    /** Used for initial separation of command word and args. */
```
Unnecessary change?

pyokagan commented on src/main/java/seedu/address/logic/parser/Parser.java:
```diff
@@ -153,23 +123,23 @@ private Command prepareAdd(String args) {
-        ArgumentTokenizer argsTokenizer = new ArgumentTokenizer(phoneNumberPrefix, emailPrefix,
-                                                                addressPrefix, tagsPrefix);
+        ArgumentTokenizer argsTokenizer =
+                new ArgumentTokenizer(PREFIX_PHONE, PREFIX_EMAIL, PREFIX_ADDRESS, PREFIX_TAG);
```
On the topic of "sneaking in style changes": don't do that if it makes the diff larger (touch more lines of code), as it makes it harder to e.g. resolve conflicts if this PR ever needs to be rebased, and also pollutes the blame history.

So, for example, https://github.com/se-edu/addressbook-level4/pull/272#discussion_r100491438 is unacceptable because it makes the diff larger than it could be.

On the other hand, these 2 lines are an acceptable "sneaky code formatting change", since even without the changing of the formatting of these lines you would still have to modify these two lines anyway.

CanIHasReview commented:
# v13

@MightyCupcakes submitted v13 for review.

* [v13 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/87d12cd6b0b7ba6558044b94099dbf72fa06d4d6)
* [v13 2/7] [Extract AddCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/61feb1c09217ebb08437ac7c0d8334ba6504081c)
* [v13 3/7] [Extract EditCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/344bf41f2fe14a166308ccf3fa6d9cc6a72645d0)
* [v13 4/7] [Extract DeleteCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/e1a2f322992d2a8f487f17b7ba7c9f41b29982d4)
* [v13 5/7] [Extract SelectCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/2d90c5fae563789101961286afc839b52408d9af)
* [v13 6/7] [Extract FindCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/6654064d64c9bc58d06b4321c483d957b2d83276)
* [v13 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/a27e2fb368a6c2bbfd598607147e30c7ba637746)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...a27e2fb368a6c2bbfd598607147e30c7ba637746)) ([:chart_with_upwards_trend: Interdiff between v12 and v13](https://github.com/se-edu/addressbook-level4/compare/1ea0df19b3308a739eb0ac8882929ad70019d731~1...1ea0df19b3308a739eb0ac8882929ad70019d731))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/13/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

MightyCupcakes commented:
Well there is no harm in standardizing everything to make it clear that it is from one single PR.

>The design can be improved by applying the Single Responsibility
Principle as follows: limit the Parser's responsibility to determining
which type of command is being parsed and move the responsibility
of parsing each type of command to a separate *CommandParser class.

There is a little disconnect between this and the following paragraph. I would modify the paragraph by appending some explanation about the bi-directional dependency like so:

>The design can be improved by applying the Single Responsibility
Principle as follows: limit the Parser's responsibility to determining
which type of command is being parsed and move the responsibility
of parsing each type of command to a separate *CommandParser class.
However, this will result in bidirectional links between these various
*CommandParser classes and the main Parser class because some
of the constants and utility methods in Parser class will be used
by the various *CommandParser classes as well.

If you have no problems, I can push out a new iteration with the new commit message right away to save on the round trip time.

damithc commented:
> There is a little disconnect between this and the following paragraph. I would modify the paragraph by appending some explanation about the bi-directional dependency like so:

Good point. Yes can do so. May be add a blank line to separate what is common to all commits from what is specific to this commit (in the spirit of 'make similar things CLEARLY similar and different things CLEARLY different' -- things that look similar but slightly different are a source of errors).

> If you have no problems, I can push out a new iteration with the new commit message right away to save on the round trip time.

Sure.

CanIHasReview commented:
# v14

@MightyCupcakes submitted v14 for review.

* [v14 1/7] [Extract common parser constants/methods to seperate classes](https://github.com/se-edu/addressbook-level4/pull/272/commits/f9cf9ade26595e3df33b1b4e9ad88d6cb6e2621b)
* [v14 2/7] [Extract AddCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/b443d880b041cabdbfdc6da7f82f7f520b1b3ef7)
* [v14 3/7] [Extract EditCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/d66527be4db0c692004c5f01530dd526fc6afc3e)
* [v14 4/7] [Extract DeleteCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/a67ac737a8a84dbcf1be1b582b9533596a49d285)
* [v14 5/7] [Extract SelectCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/03097866de219843de84193b7319f071ed597cf3)
* [v14 6/7] [Extract FindCommandParser from Parser class](https://github.com/se-edu/addressbook-level4/pull/272/commits/2658dc4f45e44cf928ca006d7f486c31eb707d92)
* [v14 7/7] [Update developer's documentation](https://github.com/se-edu/addressbook-level4/pull/272/commits/9af3aed34f48ab4fd833226db0c9a547da7bbae1)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...9af3aed34f48ab4fd833226db0c9a547da7bbae1)) ([:chart_with_upwards_trend: Interdiff between v13 and v14](https://github.com/se-edu/addressbook-level4/compare/c9f181ea7d14dc9344d81eddaa20389bb672c2ee~1...c9f181ea7d14dc9344d81eddaa20389bb672c2ee))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/272/14/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

pyokagan commented on src/main/java/seedu/address/logic/parser/AddCommandParser.java:
```diff
@@ -0,0 +1,44 @@
+     * Parses the given {@code String} of arguments in the context of the AddCommand
+     * and returns an AddCommand object for execution.
+     */
+    public Command parse(String args) {
```
Tangential to this PR: but it is unfortunate that we can't write this as:
```java
public AddCommand parse(String args) {
```
to make it explicit that the `AddCommandParser` will return an `AddCommand`. This is because we need to be able to return `IncorrectCommand` as well. So technically, this method may return either an `AddCommand` or `IncorrectCommand`, it's just not very good that this cannot be made explicit via the function signature.

We may wish to use exceptions instead in the future.

damithc replied:
Yes, let's change to use exceptions in a future PR.
